pyscf/df/autoaux.py

```diff
@@ -0,0 +1,191 @@
+#!/usr/bin/env python
+# Copyright 2024 The PySCF Developers. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+'''
+The AutoAux algorithm by ORCA
+
+Ref:
+    JCTC, 13 (2016), 554
+'''
+
+from math import factorial
+import numpy as np
+from pyscf import gto
+
+F_LAUX   = np.array([20 , 7.0, 4.0, 4.0, 3.5, 2.5, 2.0, 2.0])
+BETA_BIG = np.array([1.8, 2.0, 2.2, 2.2, 2.2, 2.3, 3.0, 3.0])
+BETA_SMALL = 1.8
+
+def _primitive_emin_emax(basis):
+    l_max = max(b[0] for b in basis)
+    emin_by_l = [1e99] * (l_max+1)
+    emax_by_l = [0] * (l_max+1)
+    e_eff_by_l = [0] * (l_max+1)
+
+    for b in basis:
+        l = b[0]
+        if isinstance(b[1], int):
+            e_c = np.array(b[2:])
+        else:
+            e_c = np.array(b[1:])
+        es = e_c[:,0]
+        emax_by_l[l] = max(es.max(), emax_by_l[l])
+        emin_by_l[l] = min(es.min(), emin_by_l[l])
+
+        cs = e_c[:,1:]
+        cs = np.einsum('pi,p->pi', cs, gto.gto_norm(l, es)) # normalize GTOs
+        cs = gto.mole._nomalize_contracted_ao(l, es, cs) # prefactors in r_ints
+        ee = es[:,None] + es
+        r_ints = gto.gaussian_int(l*2+3, ee) # \int \chi^2 r dr
+        r_exp = np.einsum('pi,pq,qi->i', cs, r_ints, cs)
+
+        k = 2**(2*l+1) * factorial(l+1)**2 / factorial(2*l+2)
+        # Eq (9) in the paper
+        #e_eff = 2 * k**2 / (np.pi * r_exp)
+        # r_exp**2 is applied in bse. For primitive functions, this leads to
+        # e_eff = exponent of the basis
```

Comment:
This was a typo in the paper; the correct expression being `r_exp**2` is obvious from the dimensional argument